src/server/hset_family.cc

```diff
@@ -1020,25 +1037,86 @@ void HSetFamily::HRandField(CmdArgList args, ConnectionContext* cntx) {
     StringVec str_vec;
 
     if (pv.Encoding() == kEncodingStrMap2) {
-      // TODO: to create real random logic.
       StringMap* string_map = (StringMap*)pv.RObjPtr();
 
-      sds key = string_map->begin()->first;
-      str_vec.emplace_back(key, sdslen(key));
+      // Randomize list
+      std::vector<dfly::StringMap::iterator> it_vec;
+      for (auto it = string_map->begin(); it != string_map->end(); ++it) {
+        it_vec.push_back(it);
+      }
+      auto rng = std::default_random_engine{};
+      std::shuffle(std::begin(it_vec), std::end(it_vec), rng);
+
+      if (count > 0) {
+        // If positive, emplace only unique elements, capped by vector size
+        for (size_t i = 0; i < std::min(size_t(count), it_vec.size()); i++) {
+          auto it = it_vec[i];
+          sds it_key = it->first;
+          sds it_value = it->second;
+          str_vec.emplace_back(it_key, sdslen(it_key));
+          if (with_values) {
+            str_vec.emplace_back(it_value, sdslen(it_value));
+          }
+        }
+      } else if (count < 0) {
+        // If negative, randomly pick elements |COUNT| times
+        for (size_t i = 0; i < size_t(-count); i++) {
+          auto it = it_vec[rand() % it_vec.size()];
+          sds it_key = it->first;
+          sds it_value = it->second;
+          str_vec.emplace_back(it_key, sdslen(it_key));
+          if (with_values) {
+            str_vec.emplace_back(it_value, sdslen(it_value));
+          }
+        }
+      } else if (it_vec.size() > 0) {
+        // Special case if count is 0, just single element in the array
+        sds it_key = it_vec[0]->first;
+        sds it_value = it_vec[0]->second;
+        str_vec.emplace_back(it_key, sdslen(it_key));
+        if (with_values) {
+          str_vec.emplace_back(it_value, sdslen(it_value));
+        }
+      }
     } else if (pv.Encoding() == kEncodingListPack) {
       uint8_t* lp = (uint8_t*)pv.RObjPtr();
       size_t lplen = lpLength(lp);
       CHECK(lplen > 0 && lplen % 2 == 0);
 
       size_t hlen = lplen / 2;
-      listpackEntry key;
-
-      lpRandomPair(lp, hlen, &key, NULL);
-      if (key.sval) {
-        str_vec.emplace_back(reinterpret_cast<char*>(key.sval), key.slen);
+      listpackEntry *keys, *values = NULL;
+      size_t lp_size = count >= 0 ? hlen : size_t(-count);
+      keys = (listpackEntry*)zmalloc(sizeof(listpackEntry) * lp_size);
+      values = (listpackEntry*)zmalloc(sizeof(listpackEntry) * lp_size);
```

Comment:
Small nit: Even if you work with C structs please use C++ style (nullptr, declaration + definition in one line, etc)